src/Pos/Delegation/Cede/Logic.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Different logic depending on "MonadCede".
+
+module Pos.Delegation.Cede.Logic
+       (
+         getPskChain
+       , getPskChainInternal
+       , detectCycleOnAddition
+       , dlgReachesIssuance
+       ) where
+
+import           Universum
+
+import           Control.Lens              (uses, (%=))
+import qualified Data.HashMap.Strict       as HM
+import qualified Data.HashSet              as HS
+
+import           Pos.Core                  (ProxySKHeavy, StakeholderId, addressHash)
+import           Pos.Crypto                (ProxySecretKey (..), PublicKey)
+import           Pos.DB                    (DBError (DBMalformed))
+import           Pos.Delegation.Cede.Class (MonadCedeRead (..))
+import           Pos.Delegation.Helpers    (isRevokePsk)
+import           Pos.Delegation.Types      (DlgMemPool)
+
+-- | Given an issuer, retrieves all certificate chains starting in
+-- issuer. This function performs a series of sequential db reads so
+-- it must be used under the shared lock.
+getPskChain
+    :: MonadCedeRead m
+    => StakeholderId -> m DlgMemPool
+getPskChain = getPskChainInternal HS.empty
+
+-- See doc for 'getPskChain'. This function also stops traversal if
+-- encounters anyone in 'toIgnore' set. This may be used to call it
+-- several times to collect a whole tree/forest, for example.
+getPskChainInternal
+    :: MonadCedeRead m
+    => HashSet StakeholderId -> StakeholderId -> m DlgMemPool
+getPskChainInternal toIgnore issuer =
+    -- State is tuple of returning mempool and "used flags" set.
+    view _1 <$> execStateT (trav issuer) (HM.empty, HS.empty)
```

Comment:
you intended to write on datatype this code. or is it another piece of code?
anyway, it'd be clearer